File: Frontend/src/services/task.js
import http from '@/services/http'
import {TASK_API} from '@/constants/api'

export const getTasks = async () => {
    const response = await http.get(TASK_API.GET_TASKS)
    return response.data
}

export const getTask = async (taskId) => {
    const response = await http.get(TASK_API.GET_TASK(taskId))
    return response.data
}

export const createTask = async (payload) => {
    const response = await http.post(TASK_API.CREATE_TASK, payload)
    return response.data
}

export const updateTask = async (taskId, payload) => {
    const response = await http.put(TASK_API.UPDATE_TASK(taskId), payload)
    return response.data
}

export const deleteTask = async (taskId) => {
    const response = await http.delete(TASK_API.DELETE_TASK(taskId))
    return response.data
}

export const getTaskById = async (taskId) => {
    const response = await http.get(TASK_API.GET_TASK(taskId))
    return response.data
}

export const getTasksByUserId = async (userId, page, per_page) => {
    const response = await http.get(TASK_API.GET_TASKS_BY_USER(userId, page, per_page))
    return response.data
}

export const getDashboardTasks = async (userId) => {
    const response = await http.get(TASK_API.GET_DASHBOARD_TASKS(userId))
    return response.data
}

export const getBarChartData = async (userId) => {
    const response = await http.get(TASK_API.GET_BAR_CHART_DATA(userId))
    return response.data
}

export const getLineChartData = async (userId) => {
    const response = await http.get(TASK_API.GET_LINE_CHART_DATA(userId))
    return response.data
}

export const getOverviewTasks = async (userId) => {
    const response = await http.get(TASK_API.GET_OVERVIEW_TASKS(userId))
    return response.data
}

File: Backend/app/services/task.py
import logging

from app.repository import TaskRepository

class TaskService:
    def __init__(self):
        self.task_repository = TaskRepository()

    def get_task_by_id(self, task_id):
        try:
            return self.task_repository.get_by_id(task_id)
        except Exception as e:
            logging.error(f"Error retrieving task by ID: {str(e)}")
            raise

    def create_task(self, data: dict):
        try:
            return self.task_repository.create(
                name=data['name'],
                description=data['description'],
                start_date=data['start_date'],
                due_date=data['due_date'],
                priority=data['priority'],
                user_id=data['user_id'],
            )
        except Exception as e:
            logging.error(f"Error creating task: {str(e)}")
            raise

    def update_task(self, task_id, data):
        try:
            return self.task_repository.update(task_id, data)
        except Exception as e:
            logging.error(f"Error updating task: {str(e)}")
            raise
    
    def delete_task(self, task_id):
        try:
            return self.task_repository.delete(task_id)
        except Exception as e:
            logging.error(f"Error deleting task: {str(e)}")
            raise
        
    def get_list_tasks(self, page=1, per_page=10):
        try:
            return self.task_repository.get_list(page, per_page)
        except Exception as e:
            logging.error(f"Error retrieving task list: {str(e)}")
            raise
        
    def get_tasks_by_user_id(self, user_id, page, per_page):
        try:
            return self.task_repository.get_list_by_user_id(user_id, page, per_page)
        except Exception as e:
            logging.error(f"Error retrieving tasks by user ID: {str(e)}")
            raise